README.md
![Version](https://img.shields.io/badge/python-3.9|3.10|3.11-brightgreen)
[![Coverage Status](https://coveralls.io/repos/github/MathisNcl/lost_cities/badge.svg?branch=master)](https://coveralls.io/github/MathisNcl/lost_cities?branch=master)

# Lost Cities

Welcome to my personal implementation of the Lost Cities card game structure. This is a very entertaining card game created by Reiner Knizia. In this game, you play one-on-one to score as many points as possible. You can play it in 5 or 10 rounds, adding up your scores for each round.

In my implementation, it's possible to play against a computer that follows personal rules.

In future i would like to set up an online interface to play with a friend or against a computer.

[Here to see latest modifications!](CHANGELOG.md)

![Game example](gameGUI.gif)

## Rules

To check the rules: [click here](https://cdn.1j1ju.com/medias/c8/66/47-lost-cities-rulebook.pdf)

## Installation

- Clone repo
- Create venv with python between 3.9 and 3.11 and `make deps`

## How to play?

For graphicqal version, run in your terminal:

```sh
play-lost-gui
```

When the game is over you can press "R" and play again!

To play in your command line, run in your terminal:

```sh
play-lost
```

Then choose your name and if you will play against a computer. Enjoy!

## Possible enhancement

- Dockerize
- Online
- Computer using reinforcement learning


CHANGELOG.md
# Changelog

## 0.2.0 - AUgust, 2023

- New gaming interface, launch with `play-lost-gui`.

## 0.1.0 - August, 2023

- First release of Lost Cities Game : only structure
- `Card`, `Player`, `ComputerPlayer` and `LostCitiesGame` objects
- 100% coverage


tests/__init__.py
"""Unit test package for lost_cities."""


tests/conftest.py
import pygame
import pytest

from lost_cities.card import Card
from lost_cities.game import LostCitiesGame
from lost_cities.gui.gui import GUIGame
from lost_cities.player import ComputerPlayer, Player


@pytest.fixture
def test_player():
    return Player("Player1", version=5)


@pytest.fixture
def computer_player():
    cp = ComputerPlayer("Test Computer Player")
    cp.hand = [
        Card("Yellow", 0),
        Card("Yellow", 0),
        Card("Yellow", 2),
        Card("Yellow", 7),
        Card("Red", 8),
        Card("Green", 0),
        Card("Green", 5),
        Card("Green", 8),
    ]

    return cp


@pytest.fixture
def game_setup():
    game = LostCitiesGame("Player1", "Player2")
    game.setup()

    return game


@pytest.fixture
def gui_game():
    pygame.init()
    game = GUIGame()
    yield game
    pygame.quit()


tests/units/test_gui.py
from lost_cities.gui.gui import GUIGame

# TODO: real tests not only launch test


def test_init(gui_game):
    assert isinstance(gui_game, GUIGame)

    assert gui_game.assets
    assert gui_game.pygame_objects
    assert gui_game.screen
    assert gui_game.font
    assert gui_game.end is False
    assert gui_game.last_action is None
    assert gui_game.running
    assert gui_game.rect_selected is None
    assert gui_game.selected_card is None
    assert gui_game.game


def test_process_without_event(gui_game):
    gui_game.can_i_play(nb_testing=5)


src/lost_cities/card.py
from typing import Any

import numpy as np
import pygame

from lost_cities.gui.utils import extract_card


class Card:
    def __init__(self, color: str, value: int, x: int = 0, y: int = 0) -> None:
        """Instanciate Card

        Args:
            color (str): color of the card
            value (int): value of the card
            x (int): x coord board
            y (int): y coord board
        """
        if color not in ["Yellow", "Blue", "White", "Green", "Red", "Purple"]:
            raise AttributeError(f"color can not be {color}")
        if value not in [i for i in range(2, 11)] and value != 0:
            raise AttributeError("value must be between 2 and 10 or a 0")
        self.color = str(color)
        self.value = value
        self.img = extract_card(self.color, self.value)
        self.surface = pygame.pixelcopy.make_surface(np.flipud(np.rot90(self.img)))
        self.x = x
        self.y = y
        self.rect = self.surface.get_rect()
        self.rect.topleft = (self.x, self.y)

    def __repr__(self) -> str:
        """Representation of a card

        Returns:
            str
        """
        return f"{self.value}:{self.color}"

    def __lt__(self, card: Any) -> bool:
        """lower than, used to sort list

        Args:
            card (Any): card to compare with

        Returns:
            bool: if this card is lower than a given card
        """
        return self.color < card.color or self.color == card.color and self.value < card.value

    def __eq__(self, card: Any) -> bool:
        """equal to

        Args:
            card (Any): card to compare with

        Returns:
            bool: if this card is equal to a given card
        """
        return self.color == card.color and self.value == card.value

    def rotate_surface_to_discard(self) -> None:
        """change surface value to rotate for discard view"""
        self.surface = pygame.pixelcopy.make_surface(np.flipud(self.img))

    def unrotate_surface_to_discard(self) -> None:
        """change surface value to unrotate for hand view"""
        self.surface = pygame.pixelcopy.make_surface(np.flipud(np.rot90(self.img)))

    def set_coord(self, x: int, y: int) -> None:
        """Set x and y coord for pygame view

        Args:
            x (int): x coord
            y (int): y coord
        """
        self.x = x
        self.y = y
        self.rect.topleft = (self.x, self.y)


src/lost_cities/game.py
import random
from typing import Any, Literal, Optional

from lost_cities import logger
from lost_cities.card import Card
from lost_cities.gui.settings import settings
from lost_cities.player import ComputerPlayer, Player


class LostCitiesGame:
    VERSION_5: list[str] = ["Yellow", "Blue", "White", "Green", "Red"]
    VERSION_6: list[str] = ["Yellow", "Blue", "White", "Green", "Red", "Purple"]

    def __init__(
        self, player1_name: str, player2_name: str, vs_computer: bool = True, version: Literal[5, 6] = 5
    ) -> None:
        """Constructor of a LostCities Game

        Args:
            player1_name (str): name of p1
            player2_name (str): name of p2
            vs_computer (bool, optional): Whether to play against computer. Defaults to True.
            version (Literal[5, 6], optional): Which version to play. Defaults to 5.
        """
        self.deck: list[Card] = []
        self.discard_piles: list[Card] = []
        self.players: list[Player] = [Player(player1_name, version)]
        self.colors: list[str] = eval(f"self.VERSION_{version}")
        if vs_computer:
            self.players.append(ComputerPlayer(player2_name, version))
        else:
            self.players.append(Player(player2_name, version))
        self.current_player: int = 0

    def setup(self) -> None:
        """Sets all cards in the deck, shuffles and gives 8 cards for each player"""
        values = list(range(2, 11))  # Values from 2 to 10
        values.extend([0] * 3)  # 0 cards
        self.deck = [Card(color, value) for color in self.colors for value in values]
        random.shuffle(self.deck)

        for _ in range(8):
            for player in self.players:
                player.hand.append(self.deck.pop())

        self.players[0].reorder_hand()

    def switch_player(self) -> None:
        """Change player cursor"""
        self.current_player = 1 - self.current_player

    def pick_card(self, chosen_pile: Optional[str] = None) -> None:
        """Pick a card either in deck either in discard

        Args:
            chosen_pile (Optional[str], optional): chosen pile by compuyrt . Defaults to None.
        """
        current_player: Player = self.players[self.current_player]
        pile: str = chosen_pile or input("Choose a pile (deck/discard): ")
        while True:
            if pile not in ["deck", "discard"]:
                logger.warning(f"{pile} is not a valid piles. Choose deck or discard.")
                pile = input("Choose a pile (deck/discard): ")
            elif pile == "discard" and len(self.discard_piles) == 0:
                logger.warning("No more card in discard, choose deck.")
                pile = "deck"
            else:
                break
        if pile == "deck":
            current_player.hand.append(self.deck.pop())
        else:
            current_player.hand.append(self.discard_piles.pop())
        current_player.reorder_hand()

    def action_play_card(self, index: Optional[str] = None, skip_card: bool = False) -> bool:
        """Method to play a valid card

        Args:
            index (Optional[str]): if not None, index to take else would be asked with input. Defaults to None.
            skip_card (bool): Whether to skip last action as taking a card and switch player. Defaults to False.

        Returns:
            bool: Whether the card has been played
        """

        current_player: Player = self.players[self.current_player]
        card_input: str = index if index is not None else input("Index card to play: ")
        while True:
            try:
                card: Card = current_player.hand[int(card_input)]
                break
            except ValueError:
                logger.warning(f"{card_input} not a int.")
                card_input = input("Index card to play: ")
            except IndexError:
                logger.warning(f"{card_input} is not reachable. Choose a valid index from 0 to 7.")
                card_input = input("Index card to play: ")

        can_play: bool = current_player.play_card(card)
        if not can_play:
            return can_play

        if skip_card is False:
            self.pick_card()
            self.switch_player()

        return can_play

    def action_discard(self, index: Optional[str] = None, skip_card: bool = False) -> None:
        """Method to discard a valid card

        Args:
            index (Optional[str]): if not None, index to take else would be asked with input. Defaults to None.
            skip_card (bool): Whether to skip last action as taking a card and switch player. Defaults to False.
        """

        current_player: Player = self.players[self.current_player]
        card_input: str = index if index is not None else input("Index card to discard: ")
        while True:
            try:
                card: Card = current_player.hand[int(card_input)]
                break
            except ValueError:
                logger.warning(f"{card_input} not a int.")
                card_input = input("Index to discard: ")
            except IndexError:
                logger.warning(f"{card_input} is not reachable. Choose a valid index from 0 to 7.")
                card_input = input("Index to discard: ")

        current_player.discard_card(card)
        card.rotate_surface_to_discard()
        x, y = settings.discard_position
        card.set_coord(x, y)
        self.discard_piles.append(card)

        if skip_card is False:
            self.pick_card()
            self.switch_player()

    def play_round(
        self, action: Optional[str] = None, index: Optional[str] = None, skip_card: bool = False, gui: bool = False
    ) -> None:
        """Play a round. Player has to choose between play or discard then pick a card in deck or discard pile

        Args:
            action (Optional[str]): if not None, action to play else would be asked with input. Defaults to None.
            index (Optional[str]): if not None, index to take else would be asked with input. Defaults to None.
            skip_card (bool): Whether to skip last action as taking a card and switch player. Defaults to False.
            gui (bool): Whether playing gui. Defaults to False.

        """
        current_player: Player = self.players[self.current_player]

        logger.info(f"{current_player.name}'s turn.")

        if isinstance(current_player, ComputerPlayer):
            action, card = current_player.choose_action()
            if action == "play":
                current_player.play_card(card)
            else:
                self.action_discard(str(current_player.hand.index(card)), skip_card=True)

            last_discarded: Optional[Card] = None
            if len(self.discard_piles) > 0:
                last_discarded = self.discard_piles[-1]

            chosen_pile: str = current_player.choose_pile(last_discarded, action)
            self.pick_card(chosen_pile)
            self.switch_player()

        else:
            logger.info(f"Your hand: {[f'{i} = {card}' for i, card in enumerate(current_player.hand)]}")
            new_action = action or input("Choose an action (play/discard): ")
            while True:
                if new_action not in ["play", "discard"]:
                    logger.warning(f"{new_action} is not a valid action. Choose play or discard.")
                    new_action = input("Choose an action (play/discard): ")
                else:
                    break
            if new_action == "play":
                while True:
                    played: bool = self.action_play_card(index=index, skip_card=skip_card)
                    if played or gui:
                        break

            else:
                self.action_discard(index=index, skip_card=skip_card)

    def play_game(self) -> None:
        """Launch the game"""
        self.setup()

        while len(self.deck) != 0:
            self.play_round()

        scores: dict[str, Any] = {player.name: player.compute_score() for player in self.players}
        logger.info(f"Scores: {scores}")
        logger.info(f"Winner: {max(scores, key=scores.get)}")  # type: ignore


def main() -> None:  # pragma: nocover
    name1: str = input("Name of Player 1? ")
    vs_computer_Yn: str = input("Play against computer? (Y/n): ")
    vs_computer: bool = True
    name2: str = "Ordi"
    if vs_computer_Yn.upper() != "Y":
        vs_computer = False
        name2 = input("Name of Player 2? ")

    game = LostCitiesGame(name1, name2, vs_computer=vs_computer)
    game.play_game()


if __name__ == "__main__":  # pragma: nocover
    main()


src/lost_cities/player.py
from typing import Any, Optional

from lost_cities import logger
from lost_cities.card import Card


class Player:
    VERSION_5: list[str] = ["Yellow", "Blue", "White", "Green", "Red"]
    VERSION_6: list[str] = ["Yellow", "Blue", "White", "Green", "Red", "Purple"]

    def __init__(self, name: str, version: int = 5) -> None:
        """Initialisation of the player

        Args:
            name (str): Name of the player
            version (int): Number of colors possible, 5 or 6. Defaults to 5.
        """
        try:
            self.name: str = name
            self.hand: list = []
            self.board: dict = {color: [] for color in eval(f"self.VERSION_{version}")}
        except AttributeError:
            raise AttributeError("version should be 5 or 6 not {version}")

    def __repr__(self) -> str:
        """Representation of the object"""
        return f"{self.name} playing with {len(self.board)} colors\nActual setup: {self.board}"

    def play_card(self, card: Card) -> bool:
        """Play a card if possible and remove it from hand

        Args:
            card (Card): Card object to play

        Returns:
            bool: Whether the card can be play or not
        """
        # check whether it is allowed
        played: bool = False
        expedition: list = self.board[card.color]
        if len(expedition) == 0 or card.value >= expedition[-1].value:
            self.hand.remove(card)
            self.board[card.color].append(card)
            played = True
            logger.info(f"{self.name} plays {str(card)}")
        else:
            logger.warning(
                f"{self.name} can not play this card {card.value}:{card.color}"
                + f" because the last card is {self.board[card.color][-1]}"
            )

        return played

    def discard_card(self, card: Card) -> Card:
        """Remove a card from the hand

        Args:
            card (Card): Card to remove

        Returns:
            Card: Card removed
        """
        self.hand.remove(card)
        logger.info(f"{self.name} discard {card}")
        return card

    def reorder_hand(self) -> None:
        """Reorder the hand"""
        self.hand.sort()

    def compute_one_score(self, expedition: list) -> int:
        """Compute score for one expedition

        Args:
            expedition (list): value in the expedition

        Returns:
            int: score of the expedition
        """
        expedition_value: int = 0

        if len(expedition) > 0:
            paris_count = sum(card.value == 0 for card in expedition) + 1
            expedition_value = sum(card.value for card in expedition) - 20

            expedition_value *= paris_count
            if len(expedition) >= 8:
                expedition_value += 20

        return expedition_value

    def compute_score(self) -> tuple[int, dict]:
        """Compute final score

        Returns:
            tuple[int, dict]: final score from your board and detail by color
        """
        total_score: int = 0
        detail: dict = {}
        for color, expedition in self.board.items():
            expedition_value = self.compute_one_score(expedition)
            total_score += expedition_value

            detail[color] = expedition_value
        return total_score, detail


class ComputerPlayer(Player):
    def __init__(self, name: str, version: int = 5) -> None:
        super().__init__(name, version)

    def __repr__(self) -> str:
        """Representation of the object"""
        return f"Computer named {self.name} playing with {len(self.board)} colors\nActual setup: {self.board}"

    def choose_action(self) -> tuple[str, Card]:
        # High importance rules
        for color in self.board:
            # Check if the player should start an expedition with a 0
            colored_hand: list[Card] = [card for card in self.hand if card.color == color]
            if (
                not self.board[color]
                and any(card.value == 0 for card in colored_hand)
                and (len(colored_hand) >= 4 or sum(card.value for card in colored_hand) >= 10)
            ):
                return ("play", Card(color, 0))

            # Play close card
            if self.board[color]:
                last_card_value: int = self.board[color][-1].value if len(self.board[color]) > 0 else 0
                closest_card: list[bool] = [card.value == last_card_value + 1 for card in colored_hand]
                close_card: list[bool] = [card.value == last_card_value + 2 for card in colored_hand]
                if any(closest_card):
                    return ("play", Card(color, last_card_value + 1))
                elif any(close_card):
                    return ("play", Card(color, last_card_value + 2))

        # Medium importance rules
        for color in self.board:
            # Discard if many card not playable
            colored_hand = [card for card in self.hand if card.color == color]
            last_value: int = self.board[color][-1].value if len(self.board[color]) > 0 else 0
            not_playable: list[Card] = [card for card in colored_hand if card.value < last_value]
            if len(not_playable) >= 3:
                return ("discard", not_playable[0])

        # Low importance
        best_move: dict[str, Any] = dict()
        for color in self.board:
            # play the closest
            colored_hand = [card for card in self.hand if card.color == color]
            last_value = self.board[color][-1].value if len(self.board[color]) > 0 else 0
            playable: list[Card] = [card for card in colored_hand if card.value >= last_value]
            potential_diff: int = playable[0].value - last_value if len(playable) > 0 else 99
            if playable and (best_move.get("card") is None or best_move.get("diff", 100) > potential_diff):
                best_move = dict(card=playable[0], diff=potential_diff)
        if best_move:
            return ("play", best_move["card"])
        else:
            return ("discard", self.hand[0])

    def choose_pile(self, discard_card: Optional[Card], last_action: str) -> str:
        """choose the best pile to take a card of

        Args:
            discard_card (Optional[Card]): last discarded card

        Returns:
            str: "deck" or "discard" choice
        """
        if discard_card is None or last_action == "discard":
            return "deck"

        board_color: list[Card] = self.board[discard_card.color]
        hand_color_cards: list[Card] = [card for card in self.hand if card.color == discard_card.color]

        if discard_card.value == 0 and not board_color and len(hand_color_cards) >= 2:
            return "discard"

        if not board_color and discard_card.value >= 4:
            return "discard"

        if board_color and discard_card.value > board_color[-1].value:
            return "discard"

        return "deck"


src/lost_cities/gui/utils.py
import numpy as np
from PIL import Image

image: np.ndarray = np.array(Image.open("src/lost_cities/gui/assets/cards.webp"))
size: tuple[int, int] = (85, 133)


def extract_card(color: str, value: int) -> np.ndarray:
    """extract card image from the original image

    Args:
        color (str): card color to extrcat
        value (int): card value to extract

    Returns:
        np.ndarray: graphical card
    """
    index_color = ["Yellow", "Blue", "White", "Green", "Red", "Back"].index(color)
    if value == 0:
        value = 11
    value -= 2

    return rgba2rgb(image[size[1] * index_color : size[1] * (index_color + 1), size[0] * value : size[0] * (value + 1)])


def rgba2rgb(rgba: np.ndarray, background: tuple[int, int, int] = (255, 255, 255)) -> np.ndarray:
    """transform a rgba image into a rgb with right opacity pixel

    Args:
        rgba (np.ndarray): rgba image (4D)
        background (tuple[int, int, int], optional): Inital background of the image rgba. Defaults to (255, 255, 255).

    Returns:
        np.ndarray: rgb image (3D)
    """
    row, col, ch = rgba.shape

    assert ch == 4, "RGBA image has 4 channels."

    rgb: np.ndarray = np.zeros((row, col, 3), dtype="float32")
    r, g, b, a = rgba[:, :, 0], rgba[:, :, 1], rgba[:, :, 2], rgba[:, :, 3]

    a = np.asarray(a, dtype="float32") / 255.0

    R, G, B = background

    rgb[:, :, 0] = r * a + (1.0 - a) * R
    rgb[:, :, 1] = g * a + (1.0 - a) * G
    rgb[:, :, 2] = b * a + (1.0 - a) * B

    return np.asarray(rgb, dtype="uint8")


src/lost_cities/gui/settings.py
from typing import Any

from pydantic_settings import BaseSettings


class Settings(BaseSettings):
    SCREEN_WIDTH: int = 1000
    SCREEN_HEIGHT: int = 800
    CARD_WIDTH: int = 85
    CARD_HEIGHT: int = 133
    BOARD_WIDTH: int = 644
    BOARD_HEIGHT: int = 183

    # END layout
    END_WIDTH: int = SCREEN_WIDTH * 3 // 4
    END_HEIGHT: int = SCREEN_HEIGHT * 3 // 4
    END_X: int = (SCREEN_WIDTH - END_WIDTH) // 2
    END_Y: int = (SCREEN_HEIGHT - END_HEIGHT) // 2

    WHITE: tuple = (255, 255, 255)
    BLACK: tuple = (0, 0, 0)
    RED: tuple = (255, 0, 0)
    board_position: tuple = (
        int(SCREEN_WIDTH / 2 - BOARD_WIDTH / 2),
        int(SCREEN_HEIGHT / 2 - BOARD_HEIGHT / 2),  # - CARD_HEIGHT
    )

    PLAYER_HEIGHT_INIT: int = board_position[1] + BOARD_HEIGHT
    COMPUTER_HEIGHT_INIT: int = board_position[1] - CARD_HEIGHT

    pile_positions: dict[str, Any] = {
        "player": {
            "Yellow": (board_position[0] + 40, PLAYER_HEIGHT_INIT),
            "Blue": (board_position[0] + 160, PLAYER_HEIGHT_INIT),
            "White": (board_position[0] + 280, PLAYER_HEIGHT_INIT),
            "Green": (board_position[0] + 405, PLAYER_HEIGHT_INIT),
            "Red": (board_position[0] + 530, PLAYER_HEIGHT_INIT),
        },
        "computer": {
            "Yellow": (board_position[0] + 40, COMPUTER_HEIGHT_INIT),
            "Blue": (board_position[0] + 160, COMPUTER_HEIGHT_INIT),
            "White": (board_position[0] + 280, COMPUTER_HEIGHT_INIT),
            "Green": (board_position[0] + 405, COMPUTER_HEIGHT_INIT),
            "Red": (board_position[0] + 530, COMPUTER_HEIGHT_INIT),
        },
    }

    deck_position: tuple = (board_position[0] + 650, board_position[1])
    deck_text_position: tuple = (int(deck_position[0] + 50), int(deck_position[1] + 25))
    discard_position: tuple = (board_position[0] + 650, board_position[1] + CARD_WIDTH + 10)

    logo_play_position: tuple = (board_position[0] - 80, board_position[1])
    discard_play_position: tuple = (board_position[0] - 80, board_position[1] + CARD_WIDTH + 10)


settings = Settings()


src/lost_cities/gui/gui.py
from typing import Any, Optional

import numpy as np
import pygame
from pygame import Rect, Surface

from lost_cities.card import Card
from lost_cities.game import LostCitiesGame
from lost_cities.gui.settings import settings
from lost_cities.gui.utils import extract_card


class GUIGame:
    def __init__(self) -> None:
        self.assets: dict[str, Surface] = {
            "board_image": pygame.image.load("src/lost_cities/gui/assets/board.webp"),
            "play_logo": pygame.image.load("src/lost_cities/gui/assets/play.png"),
            "discard_logo": pygame.image.load("src/lost_cities/gui/assets/trash.png"),
        }
        self.assets["deck"] = pygame.pixelcopy.make_surface(np.flipud(extract_card("Back", 0)))

        self.pygame_objects: dict[str, Rect] = {
            "play_logo_rect": self.assets["play_logo"].get_rect(),
            "discard_logo_rect": self.assets["discard_logo"].get_rect(),
            "deck_rect": self.assets["deck"].get_rect(),
        }

        # Update coord
        self.pygame_objects["play_logo_rect"].topleft = settings.logo_play_position  # type: ignore
        self.pygame_objects["discard_logo_rect"].topleft = settings.discard_play_position  # type: ignore
        self.pygame_objects["deck_rect"].topleft = settings.deck_position  # type: ignore

        # Pygame structure
        pygame.init()
        self.screen: Surface = pygame.display.set_mode((settings.SCREEN_WIDTH, settings.SCREEN_HEIGHT))
        pygame.display.set_caption("Lost Cities GUI")
        self.font = pygame.font.Font("src/lost_cities/gui/assets/atlantis_font.ttf", 40)
        self.running: bool = True
        self.rect_selected: Optional[pygame.Rect] = None
        self.selected_card: Optional[Card] = None
        self.last_action: Optional[str] = None
        self.end: bool = False

        # Final Game
        self.game: LostCitiesGame = LostCitiesGame("Player1", "Computer", True)
        self.game.deck = self.game.deck[:20]

    def show_setup_structure(self) -> None:
        """Blits important infos as:
        - Deck
        - Discard if there is one
        - Play and discard logo
        - Hand text
        """
        self.screen.fill(settings.WHITE)

        self.screen.blit(self.assets["deck"], settings.deck_position)
        self.screen.blit(self.font.render(f"{len(self.game.deck)}", True, settings.WHITE), settings.deck_text_position)

        # discard
        if self.game.discard_piles:
            self.game.discard_piles[-1].set_coord(settings.discard_position[0], settings.discard_position[1])
            self.screen.blit(self.game.discard_piles[-1].surface, settings.discard_position)

        # logo:
        self.screen.blit(self.assets["play_logo"], settings.logo_play_position)
        self.screen.blit(self.assets["discard_logo"], settings.discard_play_position)

        # hand
        self.screen.blit(self.font.render("Hand:", True, (200, 200, 200)), (5, 5))

    def show_played_cards(self) -> None:
        """Blits all cards played by both player in the board"""
        played_card = {"player": self.game.players[0].board, "computer": self.game.players[1].board}
        for player_side, colors in played_card.items():
            step: int = -20 if player_side == "computer" else 20
            for color, cards in colors.items():
                for i, card in enumerate(cards):
                    x, y = settings.pile_positions[player_side][color]
                    y += step * i
                    card.unrotate_surface_to_discard()
                    self.screen.blit(card.surface, (x, y))

    def show_hand(self) -> None:
        """Blits all player's cards and the rect for the selected one"""
        for i, card in enumerate(self.game.players[0].hand):
            x = 5
            y = 50 + settings.CARD_HEIGHT * 0.66 * i
            card.set_coord(x, y)
            self.screen.blit(card.surface, (x, y))

        if self.rect_selected is not None:
            pygame.draw.rect(self.screen, settings.RED, self.rect_selected, 2)

    def trigger_event(self) -> None:
        """Triggers all events based on a click"""
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                self.running = False

            elif len(self.game.deck) == 0 and event.type == pygame.KEYUP and event.unicode.lower() == "r":
                self.do_i_need_to_restart(event)

            elif self.game.current_player == 1 and len(self.game.deck) >= 1:
                self.game.play_round()

            elif event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
                self.choose_card(event)

                self.gui_action(event)

                self.pick_card_on_pile(event)

    def choose_card(self, event: pygame.event.Event) -> None:
        """Hightlight the selected card

        Args:
            event (pygame.event.Event): click event
        """
        for card in self.game.players[0].hand:
            if card.rect.collidepoint(event.pos):
                self.rect_selected = pygame.Rect(card.x, card.y, settings.CARD_WIDTH, settings.CARD_HEIGHT)
                self.selected_card = card

    def gui_action(self, event: pygame.event.Event) -> None:
        """Event action for playing and discard. Need to select a card before.

        Args:
            event (pygame.event.Event): click event
        """
        if self.selected_card is not None and len(self.game.players[0].hand) == 8:
            if self.pygame_objects["play_logo_rect"].collidepoint(event.pos):
                self.game.play_round(
                    "play", str(self.game.players[0].hand.index(self.selected_card)), skip_card=True, gui=True
                )
                self.last_action = "play"
                self.selected_card = None
                self.rect_selected = None

            elif self.pygame_objects["discard_logo_rect"].collidepoint(event.pos):
                self.game.play_round(
                    "discard", str(self.game.players[0].hand.index(self.selected_card)), skip_card=True
                )
                self.last_action = "discard"
                self.selected_card = None
                self.rect_selected = None

    def pick_card_on_pile(self, event: pygame.event.Event) -> None:
        """Either take a card from discard or deck. Only trigger if you have less than 8 cards and there is a click on
        one the two piles.

        Args:
            event (pygame.event.Event): click event
        """
        if (
            len(self.game.players[0].hand) != 8
            and self.game.discard_piles
            and self.game.discard_piles[-1].rect.collidepoint(event.pos)
            and self.last_action != "discard"
        ):
            self.game.discard_piles[-1].unrotate_surface_to_discard()
            self.game.pick_card("discard")
            self.game.switch_player()

        elif len(self.game.players[0].hand) != 8 and self.pygame_objects["deck_rect"].collidepoint(event.pos):
            self.game.pick_card("deck")
            self.game.switch_player()

    def do_i_need_to_restart(self, event: pygame.event.Event) -> None:
        """Restart the game when clicking and this is the end

        Args:
            event (pygame.event.Event): click event
        """
        if self.end is True:
            self.game = LostCitiesGame("Player1", "Computer", True)
            self.game.setup()
            self.end = False

    def stop_game(self) -> None:
        """End layout, display who wons into an ugly rectangle"""
        if len(self.game.deck) <= 0:
            self.end = True
            pygame.draw.rect(
                self.screen, settings.BLACK, (settings.END_X, settings.END_Y, settings.END_WIDTH, settings.END_HEIGHT)
            )
            scores: dict[str, Any] = {player.name: player.compute_score() for player in self.game.players}
            game_ended_text: pygame.font.Font = self.font.render(  # type: ignore
                f"Winner: {max(scores, key=scores.get)}", True, settings.WHITE  # type: ignore
            )
            score_player1: pygame.font.Font = self.font.render(  # type: ignore
                f"Player1: {scores['Player1'][0]}", True, settings.WHITE
            )
            score_computer: pygame.font.Font = self.font.render(  # type: ignore
                f"Computer: {scores['Computer'][0]}", True, settings.WHITE
            )
            # mypy says bullshits
            self.screen.blit(
                game_ended_text,  # type: ignore
                ((settings.END_WIDTH - game_ended_text.get_width()) // 2, settings.END_Y + 50),  # type: ignore
            )
            self.screen.blit(
                score_player1,  # type: ignore
                ((settings.END_WIDTH - score_player1.get_width()) // 2, settings.END_Y + 150),  # type: ignore
            )
            self.screen.blit(
                score_computer,  # type: ignore
                ((settings.END_WIDTH - score_computer.get_width()) // 2, settings.END_Y + 250),  # type: ignore
            )

    def can_i_play(self, nb_testing: Optional[int] = None) -> None:
        """Launch the game

        Args:
            nb_testing (Optional[int]): If not None, stop the game at nb_testing loop index. Defaults to None.
        """

        self.game.setup()
        cpt: int = 0
        while self.running:
            self.show_setup_structure()
            self.show_played_cards()
            self.show_hand()
            self.trigger_event()
            self.screen.blit(self.assets["board_image"], settings.board_position)
            self.stop_game()
            pygame.display.flip()
            if nb_testing is not None and nb_testing >= cpt:
                break
            cpt += 1
        pygame.quit()


def main() -> None:  # pragma: nocover
    gameGUI = GUIGame()
    gameGUI.can_i_play()


if __name__ == "__main__":  # pragma: nocover
    main()
